### PR TITLE
[broken] PD-330, ItemImage commands

### PR DESCRIPTION
**Pull request by CryptoCubik**
Cleanup of ItemImage commands and associated tests

### CODE REVIEW

File: src/api/commands/itemimage/ItemImageAddCommand.ts [hunk 1]
```diff
@@ -76,6 +81,61 @@ export class ItemImageAddCommand extends BaseCommand implements RpcCommandInterf
...
+        // No need to validate data, it's optional and a string
+        const dataId = data.params[1];
+
+        let protocol = data.params[2];
+        if (protocol) {
+            if (!dataId) {
+                // Probably impossible, but throw an error if dataId arg (the arg required before protocol) doesn't exist.
+                throw new MessageException('Protocol arg was present, but dataId arg is missing.');
+            }
+            if (typeof protocol !== 'string') {
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageAddCommand.ts]: InvalidParamException

File: src/api/commands/itemimage/ItemImageAddCommand.ts [hunk 2]
```diff
@@ -76,6 +81,61 @@ export class ItemImageAddCommand extends BaseCommand implements RpcCommandInterf
...
+                throw new MessageException('Invalid protocol.');
+            }
+            protocol = ImageDataProtocolType[protocol];
+        }
+
+        let encoding = data.params[3];
+        if (encoding) {
+            if (!protocol) {
+                // Probably impossible, but throw an error if protocol arg (the arg required before encoding) doesn't exist.
+                throw new MessageException('Encoding arg was present, but protocol arg is missing.');
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageAddCommand.ts]: its impossible, we should have allready thrown an error if protocol doesnt exists

File: src/api/commands/itemimage/ItemImageAddCommand.ts [hunk 3]
```diff
@@ -76,6 +81,61 @@ export class ItemImageAddCommand extends BaseCommand implements RpcCommandInterf
...
+            }
+            if (typeof encoding !== 'string') {
+                this.log.error('Encoding is present but not a string.');
+                throw new MessageException('Encoding is present but not a string.');
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageAddCommand.ts]: InvalidParamException

File: src/api/commands/itemimage/ItemImageAddCommand.ts [hunk 4]
```diff
@@ -76,6 +81,61 @@ export class ItemImageAddCommand extends BaseCommand implements RpcCommandInterf
...
+                throw new MessageException('Invalid encoding.');
+            }
+            encoding = ImageDataEncodingType[encoding];
+        }
+
+        // No need to validate data, it's optional and a string
+        const dataStr = data.params[4];
+        if (dataStr && !encoding) {
+            // Probably impossible, but throw an error if encoding arg (the arg required before dataStr) doesn't exist.
+            throw new MessageException('Data arg was present, but encoding arg is missing.');
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageAddCommand.ts]: impossible, exception for missing encoding should have been thrown allready

File: src/api/commands/itemimage/ItemImageListCommand.ts [hunk 5]
```diff
@@ -62,6 +52,24 @@ export class ItemImageListCommand extends BaseCommand implements RpcCommandInter
...
+        }
+
+        const idType = data.params[0];
+        if ( idType !== 'template' && idType !== 'item' ) {
+            throw new MessageException(`Invalid ID type detected <${idType}>. Expected 'template' or 'item'.`);
+        }
+
+        if (typeof data.params[1] !== 'number') {
+            this.log.error('Second arg must be numeric.');
+            throw new MessageException('Second arg must be numeric.');
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageListCommand.ts]: InvalidParamException

File: src/api/commands/itemimage/ItemImageListCommand.ts [hunk 6]
```diff
@@ -62,6 +52,24 @@ export class ItemImageListCommand extends BaseCommand implements RpcCommandInter
         }
     }
 
+    public async validate(data: RpcRequest): Promise<RpcRequest> {
+        if ( data.params.length !== 2 ) {
+            throw new MessageException('Invalid number of args. Expected 2, got <' + data.params.length + '>.');
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageListCommand.ts]: replace with MissingParamExceptions

File: src/api/commands/itemimage/ItemImageRemoveCommand.ts [hunk 7]
```diff
@@ -33,11 +33,11 @@ export class ItemImageRemoveCommand extends BaseCommand implements RpcCommandInt
...
-        if (data.params.length < 1) {
-            throw new MessageException('Requires arg: ItemImageId');
-        }
-        const itemImageModel = await this.itemImageService.findOne(data.params[0]);
-        const itemImage = itemImageModel.toJSON();
+        const itemImageId = data.params[0];
+        const itemImage = await this.itemImageService.findOne(itemImageId);
+
+        // find related itemInformation
+        const itemInformation = itemImage.related('ItemInformation').toJSON();
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageRemoveCommand.ts]: findOne returns the relations by default, theres no need to call related('ItemInformation').

File: src/api/commands/itemimage/ItemImageRemoveCommand.ts [hunk 8]
```diff
@@ -46,6 +46,22 @@ export class ItemImageRemoveCommand extends BaseCommand implements RpcCommandInt
         return this.itemImageService.destroy(data.params[0]);
     }
 
+    public async validate(data: RpcRequest): Promise<RpcRequest> {
+        if (data.params.length < 1) {
+            throw new MessageException('Requires arg: ItemImageId');
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageRemoveCommand.ts]: MissingParamException

File: src/api/commands/itemimage/ItemImageRemoveCommand.ts [hunk 9]
```diff
@@ -46,6 +46,22 @@ export class ItemImageRemoveCommand extends BaseCommand implements RpcCommandInt
...
+        }
+        // find itemImage
+        const itemImageId = data.params[0];
+        if (!itemImageId) {
```
Comment (xludx) [inline src/api/commands/itemimage/ItemImageRemoveCommand.ts]: this if is unnecessary, just validate that the id is number, and if not, throw InvalidParamException

File: test/black-box/itemimage/ItemImageAddCommand.test.ts [hunk 10]
```diff
@@ -116,7 +116,7 @@ describe('ItemImageAddCommand', () => {
         res.expectJson();
         res.expectStatusCode(404);
         expect(res.error.error.success).toBe(false);
-        expect(res.error.error.message).toBe('Invalid image protocol.');
+        expect(res.error.error.message).toBe('Invalid protocol.');
```
Comment (xludx) [inline test/black-box/itemimage/ItemImageAddCommand.test.ts]: use new Exception().getMessage()

Comment (xludx) [reply test/black-box/itemimage/ItemImageAddCommand.test.ts]: also its better not to test for res.error.error.success, you'll reach the reason for error faster.

**Comment by CryptoCubik** [comment]
Note to self: Image stuff has been rewritten but parts of this PR (i.e. command validation) are still missing from the new code

**Pull request closed without merging**